File: Python3/283. Move Zeroes.py
class Solution:
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        # count = nums.count(0) # 56ms
        # index = 0
        # for num in nums:
        #     if num:
        #         nums[index] = num
        #         index += 1
        # while index < len(nums):
        #     nums[index] = 0
        #     index += 1
        
        position = 0 # 52ms 没有显著改善
        for index, num in enumerate(nums):
            if num:
                nums[position], nums[index] = nums[index], nums[position]
                position += 1

File: Golang/283. Move Zeroes.go
// func moveZeroes(nums []int)  { // 150ms-170ms 为什么那么慢？
//     position := 0
//     for index, num := range nums {
//         if num != 0 {
//             nums[position], nums[index] = nums[index], nums[position]
//             position++
//         }
//     }
// }

func moveZeroes(nums []int)  { // 148ms
    total := len(nums)
    position := 0
    
    for _, num := range nums {
        if num != 0 {
            nums[position] = num
            position++
        }
    }
    for position < total {
        nums[position] = 0
        position++
    }
}
